Store server version in Redux (WIP for #3745)

chrisbobbe opened the pull request:
This is WIP for the part of https://github.com/zulip/zulip-mobile/issues/3745 only concerned with storing the version number in Redux.

chrisbobbe commented:
OK, I've got all tests passing; this should be ready for a first review, when it's convenient.

ray-kraesig commented on src/account/accountActions.js:
```diff
@@ -13,9 +13,10 @@ export const switchAccount = (index: number): Action => ({
 });
 
-export const realmAdd = (realm: string): Action => ({
+export const realmAdd = (realm: string, zulipVersion: string | null): Action => ({
```
This seems like it should take the entire settings object, and let the reducers pick out what they want.

(Mostly because that'll be easier to extend later. But that may be another way of saying that "what parts of the server settings object will be stored" is an implementation detail that doesn't need to be exposed here.)

chrisbobbe replied:
Sure, that reasoning makes sense. Let's ping @gnprice, though, because I think he disagrees, reasoning that we only need the zulipVersion now, and that the barrier to extending it later is fairly low.

chrisbobbe commented:
> * either way, using `number[]` instead of `string` (possibly with `rawZulipVersion` cached for reference or reporting). This keeps us from trying to compare Zulip version strings with `<` (since, incorrectly, "2.10.3" < "2.2.1").

Makes sense.

2.1.0-301-gd20ce6a

^ There's an example zulipVersion. @timabbott, I'm wondering what 301 and gd20ce6a refer to. `git log gd20ce6a` shows gd20ce6a isn't a commit ID in zulip/zulip.

timabbott commented:
Yeah, so that's the format for our versions for Git checkouts of Zulip.  That one means it's a common with 301 commits on top of Zulip 2.1.0 (with commit ID starting with `gd20ce6a`).  

@andersk is it intentional that we're not using a `2.2.x` based value in master for `zulip-git-version`?  That seems like a bug to me, the committed value (used in development) is `2.2.dev+git`.

andersk commented:
> @andersk is it intentional that we're not using a `2.2.x` based value in master for `zulip-git-version`? That seems like a bug to me, the committed value (used in development) is `2.2.dev+git`.

If we want that (and I agree that we do), we just need a tag named `2.2.dev` at 9812c6d445220c1d799ec5b856c336198ed00e30.

andersk commented:
It should be mentioned that there are already many possible `git describe` output formats, so you shouldn’t make too many assumptions:

* `2.1.0-rc1`
* `2.1.0-rc1-112-g35959d43c4`
* `2.1.0`
* `2.1.1-50-gd452ad31e0`
* `2.2-dev-426-gdf6b90db3c` (if we make a `2.2-dev` tag)

timabbott commented:
I think the 2.2-dev tag is worth doing.  @andersk can you push the tag (and also update the release checklist to mention doing that?)

andersk commented:
> I think the 2.2-dev tag is worth doing. @andersk can you push the tag (and also update the release checklist to mention doing that?)

Done.

ray-kraesig commented:
> It should be mentioned that there are already many possible `git describe` output formats, so you shouldn’t make too many assumptions:
> 
> * `2.1.0-rc1`
> * `2.1.0-rc1-112-g35959d43c4`
> * `2.1.0`
> * `2.1.1-50-gd452ad31e0`
> * `2.2-dev-426-gdf6b90db3c` (if we make a `2.2-dev` tag)

Assuming `dev` < `rc0` < `rc1` < ... < ø (does it?), we can still use the lexicographic ordering on `number[]` with all of these if `dev` maps to `[-2]` and `rcN` maps to `[-1,N]`.

I think all we really *need*, though, is major-minor[-patch]; anything past that is gravy.

ray-kraesig commented:
Given the amount of discussion that's enused on this topic (see also [discussion on CZO](https://chat.zulip.org/#narrow/stream/243-mobile-team/topic/Server.20version.20number/near/814933)), I should probably note: while we'll want a parsed and partial-ordered server-version at some point, it doesn't have to block this PR. An unparsed version-string will still be useful for #3745.

chrisbobbe commented:
(Not yet ready for review; I just remembered I haven't addressed Ray's comments on the earlier commits.)

chrisbobbe commented:
> alternately, consider using zulipVersion: string instead. This requires us to write a migration, but Zulip servers have been sending zulip_version since at least 1.5.1 in 2017. (Not sure if that's old enough not to worry about.)

As @gnprice mentioned to me on Friday, such a migration would have to exist outside the pattern of our current migrations because it requires making an API call. I think we instead landed on it being OK to allow zulipVersion to be empty, since most of the places where we'd need it are after the user is logged in and REALM_INIT has been dispatched. I'm fine to make the "empty" state be undefined instead of null, as you've suggested; I agree with your reasoning.

chrisbobbe commented:
OK, this is ready for another round of review. I've added version parsing code with tests, and addressed Ray's earlier feedback.

gnprice commented on src/__tests__/exampleData.js:
```diff
@@ -90,16 +90,29 @@ export const makeCrossRealmBot = (args: { name?: string } = {}): CrossRealmBot =
+  shouldHaveApiKey?: boolean,
+  shouldHaveZulipVersion?: boolean,
+}): Account => {
+  const { email, realm, shouldHaveApiKey = false, shouldHaveZulipVersion = false } = args;
```
One thing that helps keep the code in the test cases themselves simple is when they can skip mentioning options for things that are boring for that test case, and only mention something like the API key when the point of the test case is specifically related to whether the API key is present or absent.

It looks like in the current version, this works out very nicely for `shouldHaveZulipVersion` -- it's actually not mentioned in any call sites except `selfAccount`.

For `shouldHaveApiKey`, all but a few of the call sites are passing `true`. That's a sign that that should be the default -- most of those test cases aren't really thinking about the API key, and it's just a background assumption of a lot of our code that there is an API key in the active account.

ray-kraesig commented:
> As @gnprice mentioned to me on Friday, such a migration would have to exist outside the pattern of our current migrations because it requires making an API call.

For the record: I was thinking the migration could just use "0.0" or similar. (An honest `undefined` is probably a better idea all around, though.)

chrisbobbe commented on src/__tests__/exampleData.js:
```diff
@@ -90,16 +90,29 @@ export const makeCrossRealmBot = (args: { name?: string } = {}): CrossRealmBot =
 export const selfUser: User = makeUser({ name: 'self' });
-export const selfAccount: Account = makeAccount(selfUser);
+export const selfAccount: Account = makeAccount({
+  email: 'self@example.org',
```
OK, makes sense! Would you say makeAccount should take these args, then? I checked, and it seems that realm isn't also covered by the User type.

```
args: {
  user: User,
  realm: string,
  shouldHaveApiKey?: boolean,
  shouldHaveZulipVersion?: boolean,
}
```

chrisbobbe replied:
On REALM_ADD, an Account object with an empty string for `email` is added. I'll need to simulate that somehow, and if I get the email from a User object, it'll always be non-empty; even if I pass an empty `name`, `email` ends up being '@example.org'.

chrisbobbe commented on src/utils/__tests__/version-test.js:
```diff
@@ -0,0 +1,74 @@
+
+describe('ZulipVersion.prototype.raw()', () => {
+  const raw = randString();
+  test(`new ZulipVersion(${raw}).raw() === ${raw}`, () => {
```
Makes sense. Do you think I shouldn't use the names `${versions[i][k].raw()} is at least ${versions[j][l].raw()}` for the .isAtLeast tests? I suppose the names will still be the same run to run, there's just a very large number of them.

gnprice replied:
Yeah, no, I think that pattern is great :-) If a test fails, that helps make it completely clear what the test was about.

chrisbobbe commented:
OK, I think I've addressed all of your comments, @gnprice, but I ended up doing some fairly substantial reorganization of the commits, so hopefully I haven't introduced more issues. But I'm hoping at least it's easier to review with this organization. :)

One thing to note in this revision is that I introduced shouldHaveZulipVersion and defaulted it to false, even though you've rightly pointed out that it should default to true. In later commits that need to test with a zulipVersion present, I switched it to default to true. This back and forth is intended to reduce the size of the patch, to help make reviewing easier; I figured it would be more lines of code if I switched a lot of the callers back and forth like this.

chrisbobbe commented on src/__tests__/exampleData.js:
```diff
@@ -90,16 +90,43 @@ export const makeCrossRealmBot = (args: { name?: string } = {}): CrossRealmBot =
+  return deepFreeze({
+    realm,
+    email,
+    apiKey: shouldHaveApiKey ? apiKey : '',
```
You mean `apiKey = ''`, right? (And keep `apiKey?: string` as-is but remove `shouldHaveApiKey`.)

chrisbobbe replied:
Just a note that this means if the caller wants an API key, they have to provide it themselves, e.g., by using randString() or '123'.

chrisbobbe replied:
Also, the same pattern is used in a later commit with the zulipVersion; perhaps this should be changed as well.

gnprice replied:
Ah, I meant perhaps the reverse of that -- I did mean `apiKey: ''`, and by that I meant (and didn't say this clearly) that's what the *caller* would say.

So:
* signature has `apiKey?: string`
* when it's not provided, the default is to make a random one (just like in master)

... OK, and you just asked me about this in person :slightly_smiling_face: and we discussed, so I'll leave this off here.

chrisbobbe commented on src/api/registerForEvents.js:
```diff
@@ -18,7 +19,7 @@ type RegisterForEventsParams = {|
 
 /** See https://zulipchat.com/api/register-queue */
-export default (auth: Auth, params: RegisterForEventsParams) =>
+export default async (auth: Auth, params: RegisterForEventsParams): Promise<InitialData> =>
```
Yep :) I just updated the commit message with the detail about exported values.

chrisbobbe commented on src/actionTypes.js:
```diff
@@ -181,6 +182,7 @@ type LogoutAction = {|
 type RealmInitAction = {|
   type: typeof REALM_INIT,
   data: InitialData,
+  zulipVersion?: string,
```
Ah. So, while it seems like we should have zulipVersion be optional on Account, it should be required in the Redux actions. (Indeed, it isn't and hasn't been optional in ApiResponseServerSettings.)

chrisbobbe commented on src/types.js:
```diff
@@ -78,6 +78,11 @@ export type InputSelection = {|
   ...Auth,
 
+  /**
+   * To determine behavior of the mobile app, and for Sentry reports.
```
Haha yes, it looks quite jarring as a description of intent, vs. abilities provided. I was optimizing for accuracy on the latter, but intent makes much more sense given the preference for "why" vs. "what" comments.

chrisbobbe commented on src/utils/__tests__/version-test.js:
```diff
@@ -0,0 +1,74 @@
+        for (let l = 0; l < versions[j].length; l++) {
+          test(`${versions[i][k].raw()} is at least ${versions[j][l].raw()}`, () => {
+            // Compare versions[i][k] to versions[j][l] for all valid i, j, k, l
+            expect(versions[i][k].isAtLeast(versions[j][l])).toBe(i >= j);
```
> Better yet, though, you could make the test name be completely descriptive of the test: pull the i >= j check outside, and use it to decide both the test name and whether the value is expected to be true or false.

I had this in a previous iteration but I must have lost it in a refactor; thanks for pointing this out!

chrisbobbe commented:
OK, ready for another review!

gnprice commented on src/utils/zulipVersion.js:
```diff
@@ -0,0 +1,149 @@
+export class ZulipVersion {
+  _raw: string & JSONable; // how to mark $ReadOnly with Flow?
+  _comparisonArray: number[];
+  _loggingArray: number[] & JSONable;
```
Do these `& JSONable` have any effect? I think `string` and `number[]` should be subtypes of `JSONable` already.

gnprice replied:
(And I don't think it's helpful to the human reader to redundantly say that here: it's a basic fact of JSON that it cleanly handles strings, numbers, and arrays of them.)

chrisbobbe replied:
True. I had left a code comment explaining that this was done for the reader's benefit, to whitelist the fields that could dependably be sent to Sentry, but this is better done in the JSDoc.

gnprice commented on src/account/accountsReducer.js:
```diff
@@ -27,11 +32,20 @@ const realmAdd = (state, action) => {
 
+const realmInit = (state, action) => [
+  {
+    ...state[0], // Assume state[0] is the active account
```
Added some jsdoc in 8bf489397, plus filled in some glossary entries in the two commits before that.

chrisbobbe replied:
> Added some jsdoc in 8bf4893, plus filled in some glossary entries in the two commits before that.

Great, thanks!

> (Or of encapsulating it so less code cares, or design a better structure in the first place.)

You've mentioned this before; I made a note of it at https://github.com/zulip/zulip-mobile/issues/2380#issuecomment-588578501. Perhaps not a top priority, but that issue might have some helpful context.

chrisbobbe commented:
OK, I believe I've addressed all of your comments now. I've also used the quantifier `{4,40}` instead of `+` to match the commit ID for `commitsMatch`. I considered adding a $ at the end, to discard the commit info if it's followed by garbage, but I decided against it; the garbage itself can be ignored, while preserving the commit info. This assumes any garbage wouldn't suggest a false positive for the commit info; I suppose `{4,40}` might make this a bit less likely.

chrisbobbe commented:
> and there's this from last round (copying here so I don't forget about it)

Ohhhh, I saw your comment saying you’d gone ahead and added a jsdoc, in https://github.com/zulip/zulip-mobile/pull/3839#discussion_r384221034, and my eyes skimmed right past the fact that you were talking about two completely different jsdocs. Oops! We’ll go over this tomorrow. 🙂

gnprice commented:
And merged. (After the edits we made together just now in ZulipVersion.js, mainly to jsdoc and comments.) Thanks again!
